<a href="https://colab.research.google.com/github/aylenalderete/Machine-Learning-Carrera/blob/main/Technical_Challenge_customers_and_orders.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Technical Challenge ! Custiomers and Orders - The lifeblood of any business!

This notebook depends on two data files: customers.scv and orders.csv. These files are hosted in the cloud (or Github), so for simplicity, we are providing you with the code to download and save the two files in a folder called 'sample_data' which lives within this notebook session (not on your local computer nor Google Drive).

![sample_data folder in notebook session](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/sample_data_folder.png)

Please execute the first code snippet before moving to the rest of the exercise as this code will import the data from the files for you to be able to execute the rest of the exercise.

Remember as we said in the previous lecture, you have to save this notebook into your Google Drive by going to menu File/Save a Copy In Drive, or clicking in the button "Copy to Drive". This way, you will not lose any work you've done and the file will persist with your latest changes in your Google Drive. Also we recommend you to rename your .ipynb in your Google Drive, so you can easily find it in the future.

In [ ]:
import requests

def import_data_files():
  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/customers.csv')
  with open('./sample_data/customers.csv', 'wb') as f:
    f.write(r.content)

  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv')
  with open('./sample_data/orders.csv', 'wb') as f:
    f.write(r.content)
  
import_data_files()
print("Customers and orders CSV files have been added './sample_data'")

Customers and orders CSV files have been added './sample_data'


# Exercise 1: Processing Customers data (difficulty medium)

The sample customer data in 'customers.csv' file has just 5 columns: CustomerId, First Name, Last Name, City and State

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/customers.png)

We strongly recommend that you complete from the [Prep Course: Intro to Python](https://colab.research.google.com/github/anyoneai/notebooks/blob/main/python3_crash_course.ipynb) the following section:
- Section 7: File I/O (to understand how to read CSV file) 
- Section 6: For Loop (to navigate the contents of the CSV file)
- Section 5: Tuples, Lists, and Dictionaries (to manipulate the data of the CSV file)

With this, we hope you can complete this exercise successfully. Although if you want to solve this with libraries or any other way, you are welcome to do it your way.

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://github.com/anyoneai/notebooks/blob/main/customers_and_orders/data/customers.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, keep in mind that data might not be clean and you might have to figure out how to deal with that data from the code, without having to modify the data source.

Below are the 5 questions you'll have to answer to pass the evaluation:

**Question 1:** How many customers are in the file?
(as help, we have added some comments and starter code to help you structure the solution)

In [ ]:
from os.path import exists
import csv

# check if file can be found and open the file
datafile = "./sample_data/customers.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

# customers file is in the sample_data folder and can be opened.
# Using csv reader is one way of dealing with the file 
# but it's up to you how you want to proceed
with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')

  # loop through the rows and count etc. THE REST IS UP TO YOU!
  customers_size = 0
  for customer in csvreader:
    if customer[0]:
      customers_size += 1
  print("There are " + str(customers_size) + " customers")

print("All done!")

There are 603 customers
All done!


**Answer 1: There are 603 customers.**

**Question 2:** In how many different states do the customers live in?

In [ ]:
# open file and init vars
# evaluate the contents of the file
datafile = "./sample_data/customers.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')

  states = []
  for state in csvreader:
    if state[4].upper().replace(" ", "") not in states and state[4] != 'State':
      states.append(state[4])

  print("Customers live in " + str(len(states)) + " different states")

Customers live in 14 different states


**Answer 2: customers live in 14 different states**

**Question 3** What is the state with most customers?

In [ ]:
# open file and init vars
# evaluate the contents of the file
datafile = "./sample_data/customers.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')
  
  states = {}

  for state in csvreader:
    state_clean = state[4].upper().replace(" ", "")
    if state_clean not in states and state_clean != 'STATE':
      states[state_clean] = 1
    elif state_clean in states:
      states[state_clean] += 1

  max_value = max(states.values())
  max_keys = [k for k, v in states.items() if v == max_value]

  print("{} with {} is the state with most customers".format(max_keys, max_value))

['CA'] with 569 is the state with most customers


**Answer 3: CA is the state with most customers.**

**Question 4** What is the state with the least customers?

In [ ]:
# open file and init vars
# evaluate the contents of the file
datafile = "./sample_data/customers.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')
  
  states = {}

  for state in csvreader:
    state_clean = state[4].upper().replace(" ", "")
    if state_clean not in states and state_clean != 'STATE':
      states[state_clean] = 1
    elif state_clean in states:
      states[state_clean] += 1
  
  min_value = min(states.values())
  min_keys = [k for k, v in states.items() if v == min_value]

  print("These are the states: " + str(states))
  print("ANSWER: {} with the value of {} are the states with least customers".format(min_keys, min_value))

These are the states: {'CA': 569, 'AZ': 6, 'NV': 8, 'FL': 3, 'WA': 1, 'NH': 1, 'ID': 1, 'CO': 3, 'TX': 2, 'NM': 3, 'OR': 1, 'UT': 2, 'MA': 1, 'IN': 1}
ANSWER: ['WA', 'NH', 'ID', 'OR', 'MA', 'IN'] with the value of 1 are the states with least customers


Answer 4: 'WA', 'NH', 'ID', 'OR', 'MA', 'IN' are the states with least customers

**Question 5:** What is the most common last name?

In [ ]:
# open file and init vars
# evaluate the contents of the file
datafile = "./sample_data/customers.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')

  names = {}

  for name in csvreader:
    name_clean = name[2].upper().replace(" ", "")
    if name_clean not in names and name_clean != 'LASTNAME':
      names[name_clean] = 1
    elif name_clean in names:
      names[name_clean] += 1

  max_value = max(names.values())
  max_keys = [k for k, v in names.items() if v == max_value]
  print("These are the names: " + str(names))
  print("ANSWER: {} with the value of {} is the most common last name".format(max_keys, max_value))

These are the names: {'PIEDRAHITA': 1, 'ABRAHAM': 2, 'JAUREGUI': 2, 'MOTA-SANCHEZ': 2, 'MCCRORIE': 1, 'RAVELO': 1, 'KLINK': 1, 'SMITH': 8, 'TAN': 1, 'HUYNH': 4, 'GILARTELLI': 3, 'PERSKY': 1, 'MCCARLEY': 2, 'GANTOUS': 2, 'WOLLMAN': 3, 'TUELL': 2, 'COSIO': 1, 'SOMOSOT': 1, 'QUIJADA': 2, 'MARIN': 3, 'PARK': 1, 'LIGHTFOOT': 1, 'MCMURRAY': 3, 'KIRSCHNER': 1, 'ZHANG': 3, 'STINSON': 2, 'GALVAN': 1, 'MEGERDICHIAN': 2, 'ROCHA': 4, 'MUSLEH': 2, 'SLOAN': 2, 'ZAPATA': 1, 'ROELANTS': 1, 'VOJNAR': 1, 'CHO': 2, 'HAROLD': 1, 'REYES': 4, 'THOMPSON': 2, 'SCATESJOHNSON': 2, 'MORRIS': 1, 'CAMPOS': 1, 'HARRIS': 3, 'TZARAX': 1, 'COHN': 1, 'MARTINEZ': 2, 'SULLIVAN': 1, 'CARREON': 1, 'MCMAHON': 4, 'MONTES': 2, 'MOPSIKOFF': 1, 'TUDELA': 1, 'GARCIA': 4, 'COFFEY': 2, 'KIM': 1, 'GONZALEZ': 4, 'JOHNSON': 4, 'BANIK': 2, 'VALEN': 1, 'HATHEWAY': 1, 'NEWHOUSE': 1, 'PADILLA': 1, 'SPILLMAN': 1, 'HILLARD': 1, 'BROVKO': 1, 'GOMEZ': 5, 'ROBERTSON': 2, 'HODGSON': 1, 'AMBULO': 2, 'SHRESTHA': 2, 'WELSH': 1, 'OJEDA': 1, 'HALVE

Answer 5: 'SMITH' with the value of 8 is the most common last name

# Exercise 2: Processing Orders data (difficulty high)

The second sample files contains orders placed by customers from the first file. Be careful, this file has many rows and you most likely should not print the contents of the whole file. 

The file contains the following columns: CustomerID, OrderID, Date, OrderTotal, ProductName, Price

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/orders.png)

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, the data is not clean and you will have to figure out how to deal with that data from the code, without having to modify the data source.



**Question #1:** How many unique orders are in the orders.csv file?

**Question #2:** What is the average number of items per order (rounded to two decimal places)?

**Question #3:** What is the highest number of items per order?

**Question #4:** What is the number of orders placed in October 2021?

**Question #5:** Which customer spent the most amount of money in 2021?

**Question #6:** Historically, what is the best month for sales?

Once you get your answers, remember to go back to the course and introduce them in the multiple choice quiz

In [ ]:
# open and evaluate orders.csv
import math

datafile = "./sample_data/orders.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')

  orders = {}
  items_per_order = {}
  orders_date = {"2021-10": 0}
  orders_2021 = {}
  best_months = {}

  for order in csvreader:
    order_clean = order[1].upper().replace(" ", "")
    items = order[1].upper().replace(" ", "")
    order_date = "202110" in str(order[2].replace("-", ""))
    order_2021 = "2021" in str(order[2].replace("-", ""))
    months = order[2][5:7]

    if order_clean not in orders and order_clean != 'ORDERID':
      orders[order_clean] = 1
    elif order_clean in orders:
      orders[order_clean] += 1

    if items not in items_per_order and items != 'ORDERID':
      items_per_order[items] = 1
    elif items in items_per_order:
      items_per_order[items] += 1

    if order_date:
      orders_date["2021-10"] += 1

    if order_2021:
      if order[0] in orders_2021.keys():
        orders_2021[order[0]] = abs(float(orders_2021[order[0]])) + abs(float(order[5]))
      else:
        orders_2021[order[0]] = order[5]
    
    if months not in best_months:
      best_months[months] = 1
    else:
      best_months[months] += 1


  best_month = max(best_months.values())
  max_mount = sorted(orders_2021.values(), key=float)
  key_mount = 0

  for key, value in orders_2021.items():
    if max_mount[-1] == value:
      key_mount = key


  min_value = min(orders.values())
  min_keys = [k for k, v in orders.items() if v == min_value]
  
  # print("ANSWER 1: there are {} people who made a unique ({}) order".format(len(min_keys), min_value))
  print("ANSWER 1: there are {} unique orders".format(len(orders)))
  print("ANSWER 2: the average number of items per order is " + str(round(sum(items_per_order.values()) / len(items_per_order.values()), 2)))
  print("ANSWER 3: the highest number of items per order is {}".format(max(items_per_order.values())))
  print("ANSWER 4: the number of orders placed in October 2021 is {}".format(orders_date.values()))
  print("ANSWER 5: the CustomerID of the customer who spent the most amount of money (${}) in 2021 is {}".format(max_mount[-1], key_mount))
  print("ANSWER 6: the best month for sales is January with {} sales.".format(best_month))



ANSWER 1: there are 16672 unique orders
ANSWER 2: the average number of items per order is 1.76
ANSWER 3: the highest number of items per order is 35
ANSWER 4: the number of orders placed in October 2021 is dict_values([437])
ANSWER 5: the CustomerID of the customer who spent the most amount of money ($7675.0) in 2021 is 5172443
ANSWER 6: the best month for sales is January with 2876 sales.


# Finished!

Hope this was not too difficult and slicing and dicing the datasets was some fun. Now head on back to the course and provide the answers to the questions from this exercise.